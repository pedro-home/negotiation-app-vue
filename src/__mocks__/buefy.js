import Buefy from 'buefy'

export default {
  ...Buefy,
  install: function(Vue, config) {
    Buefy.install(Vue, {
      ...config,
      defaultContainerElement: 'body',
    })
  },
}
