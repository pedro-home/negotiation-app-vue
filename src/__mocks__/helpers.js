import { mount } from '@vue/test-utils'

export const resetDOM = () => {
  document.head.innerHTML = ''
  document.body.innerHTML = ''
}

export const render = (Component, props) => {
  resetDOM()

  return mount(Component, {
    propsData: props,
    attachToDocument: true,
  })
}
