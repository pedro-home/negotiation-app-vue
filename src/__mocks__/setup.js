import Vue from 'vue'
import Buefy from 'buefy'

// Makes sure that Jest uses buefy
Vue.use(Buefy)
