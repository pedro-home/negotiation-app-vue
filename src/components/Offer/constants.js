/**
 * Number regex to validate if a value is in a number format
 */
export const VALIDATE_NUMBER_REGEX = /^-?\d+(,\d+)*(\.\d+(e\d+)?)?$/
