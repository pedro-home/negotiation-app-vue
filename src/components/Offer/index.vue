<template>
  <div>
    <b-field>
      <b-input v-model="inputValue" :placeholder="text" @input="validate"></b-input>
    </b-field>
    <b-button :disabled="isDisabled" @click="() => onSubmit(inputValue)">Submit</b-button>
  </div>
</template>

<script>
import { VALIDATE_NUMBER_REGEX } from './constants'

export default {
  props: {
    text: {
      type: String,
      required: true,
    },
    onSubmit: {
      type: Function,
      required: true,
    },
  },
  methods: {
    validate: function() {
      // Enables submit button according to the input value validation
      this.isDisabled = !this.inputValue.match(VALIDATE_NUMBER_REGEX)
    },
  },
  data: () => ({
    inputValue: '',
    isDisabled: true,
  }),
}
</script>
