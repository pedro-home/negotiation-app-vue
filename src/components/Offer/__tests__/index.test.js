import Vue from 'vue'
import { render } from '@/__mocks__/helpers'
import Offer from '../index.vue'

describe('Given <Offer />', () => {
  let wrapper
  let props = {}
  beforeEach(() => {
    const defaultProps = { onSubmit: () => {}, text: '' }
    wrapper = render(Offer, { ...defaultProps, ...props })
  })

  describe('When initializes', () => {
    describe('And uses default props', () => {
      it('Then should match snapshot', () => {
        expect(wrapper.element).toMatchSnapshot()
      })
    })

    describe('And uses onSubmit prop', () => {
      let handleSubmit
      beforeAll(() => {
        handleSubmit = jest.fn()
        props = { onSubmit: handleSubmit }
      })

      it('Then should get a click event', async () => {
        wrapper.setData({ isDisabled: false })
        await Vue.nextTick()

        const button = wrapper.find('button')
        button.trigger('click')
        await Vue.nextTick()

        expect(handleSubmit).toHaveBeenCalledTimes(1)
      })
    })

    describe('And uses text prop', () => {
      beforeAll(() => {
        props = { text: 'Testing' }
      })

      it('Then should match snapshot', () => {
        expect(wrapper.element).toMatchSnapshot()
      })

      it('And should render a placeholder', () => {
        expect(
          wrapper.find('input').element.getAttribute('placeholder')
        ).toEqual('Testing')
      })
    })

    describe('And applies incorrect value into input', () => {
      beforeEach(async () => {
        wrapper.find('input').setValue('Test')
        await Vue.nextTick()
      })
      it('Then should disable submit button', () => {
        expect(wrapper.find('button').element.getAttribute('disabled')).toBe(
          'disabled'
        )
      })
    })

    describe('And applies correct value into input', () => {
      beforeEach(async () => {
        wrapper.find('input').setValue(10)
        await Vue.nextTick()
      })
      it('Then should enable submit button', () => {
        expect(
          wrapper.find('button').element.getAttribute('disabled')
        ).toBeNull()
      })
    })
  })
})
