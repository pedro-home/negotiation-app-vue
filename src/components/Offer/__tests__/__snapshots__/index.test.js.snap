// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given <Offer /> When initializes And uses default props Then should match snapshot 1`] = `
<div>
  <div
    class="field"
  >
    <!---->
     
    <div
      class="control is-clearfix"
    >
      <input
        autocomplete="on"
        class="input"
        placeholder=""
        type="text"
      />
       
      <!---->
       
      <!---->
       
      <!---->
    </div>
     
    <!---->
  </div>
   
  <button
    class="button"
    disabled="disabled"
    type="button"
  >
    <!---->
     
    <span>
      Submit
    </span>
     
    <!---->
  </button>
</div>
`;

exports[`Given <Offer /> When initializes And uses text prop Then should match snapshot 1`] = `
<div>
  <div
    class="field"
  >
    <!---->
     
    <div
      class="control is-clearfix"
    >
      <input
        autocomplete="on"
        class="input"
        placeholder="Testing"
        type="text"
      />
       
      <!---->
       
      <!---->
       
      <!---->
    </div>
     
    <!---->
  </div>
   
  <button
    class="button"
    disabled="disabled"
    type="button"
  >
    <!---->
     
    <span>
      Submit
    </span>
     
    <!---->
  </button>
</div>
`;
